Add libsecp256k1 & libsecp256k1-py-bindings & coincurve

MementoRC opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

MementoRC commented:
@conda-forge/help-python-c, I think I need help, the windows build fails on python-3.8, I don't understand the `dependency-links.txt` error message. It worked yesterday on a previous submission, but I was to ashamed of the 300+ commits I went through to submit it. This is my first recipe and the learning curve is brutal for this old of a person.
Thanks for the help

carterbox commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,94 @@
+
+extra:
+  recipe-maintainers:
+    - https://github.com/ofek
+    - https://gihub.com/MementoRC
```
Just the GitHub usernames, not urls.

MementoRC replied:
Will do

carterbox commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,94 @@
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: c86626afe417a09d8e80e56780efcae3ae516203b23b5ade84813916e1c94fc1
```
Please use the GitHub release instead of the PYPI souce dist so that you don't have to manually add the license files here. That way, if they change in the future release, they will automatically be updated.

MementoRC replied:
Should I systematically prefer github releases (I misinterpreted the doc thinking Pypi was preferred). How about a particular commit? I see you have done that in one of your recipes (I am trying to find best practice )?

carterbox replied:
You can use the PYPI source dists if it includes all of the necessary artifacts, but in this case it doesn't. The developers of this package didn't think to include the license files in the PYPI source dist.

MementoRC replied:
Updated to using github commit. Resolved

carterbox commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,94 @@
+  summary: Cross-platform Python CFFI bindings for libsecp256k1
+  doc_url: https://ofek.dev/coincurve/
+  description: |
+    This library provides well-tested Python bindings for libsecp256k1, 
```
It looks like this package is redistributing https://github.com/bitcoin-core/secp256k1. This is not allowed. secp256k1 should be build as a separate conda package and dynamically linked.

This package seems to have support for using system installed secp256k1.
https://github.com/ofek/coincurve/blob/2dfea673835e524ffaeb11dc1c2aecaf23188195/setup_support.py#L74C8-L74C8

carterbox replied:
> What is the typical merge time for a PR.

Reviewers are unpaid volunteers, so there is not typical amount of time.

> but only import the minimum from the libsecp256k1 recipe.

I don't understand what you mean.

MementoRC replied:
> Reviewers are unpaid volunteers, so there is not typical amount of time.
Of course, I just wanted to understand what strategy I should use bundling or not, no other intent. Another way to ask is as a reviewer, do you prefer to review a bundle of 5+ packages or singled-out recipes?

> I don't understand what you mean.
The part of `libsecp256k1` that I included was a modified version that only had the parts that I needed for `coincurve` to build successfully on all the platforms

So the plan forward is to merge `libsecp256k1` to this recipe, do we agree?
I moved BIP to work in progress and will wait until `coincurve` is merged?

carterbox replied:
> Of course, I just wanted to understand what strategy I should use bundling or not, no other intent. Another way to ask is as a reviewer, do you prefer to review a bundle of 5+ packages or singled-out recipes?

Packages should be bundled into related groups of around 5-ish packages or less.

> The part of libsecp256k1 that I included was a modified version that only had the parts that I needed for coincurve to build successfully on all the platforms So the plan forward is to merge libsecp256k1 to this recipe, do we agree?

This single PR should contain multiple recipes such that libsecp256k1 is separate from coincurve. Coincurve recipe can depend on the output of the libsecp256k1 recipe.

> I moved BIP to work in progress and will wait until coincurve is merged?

Perhaps wait until your first PR is merged before working on the others, so that you have more experience.

MementoRC replied:
Separated `libsecp256k1`. Resolved

carterbox commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,94 @@
+    - requests
+    - setuptools
+    - wheel
+    - autoconf  # [not win]
+    - automake  # [not win]
+    - libtool  # [not win]
+    - m4  # [not win]
+    - make  # [not win]
+    - pkg-config  # [not win]
+    - asn1crypto  # [win]
+    - m2-bash  # [win]
+    - m2-autoconf  # [win]
+    - m2-automake-wrapper  # [win]
+    - m2-gzip  # [win]
+    - m2-libtool  # [win]
+    - m2-m4  # [win]
+    - m2-pkg-config  # [win]
+    - m2-sed  # [win]
+    - m2w64-gcc  # [win]
+    - m2w64-make  # [win]
+    - posix  # [win]
+    - pydflow  # [win]
```
I would suggest using msvc toolchain instead of m2 if you have the option. The m2 conda toolchain isn't maintained well.

MementoRC replied:
Could you point me to documentation on the `msvc` alternative, is it what is referred to as `msys2`. This is only for personal education, based on your review, it looks like the latter work I did on coincurve is preferred.

Would you advise to close this PR?

carterbox replied:
By MSVC, I mean Microsoft Visual Studio which is the default compiler for the Windows platform. The Visual Studio compiler is provided by the `{{ compiler('c') }}` package. Whether or not this project will allow you to choose different compilers is up to the project.

carterbox replied:
> I do use MSVC, but I need all these in order to make the old make-based build work.

OK.

> You're moving too fast for me, I don't want to waste your time, but I also want to understand what I do wrong. You closed the PR that I thought was the better one, so I am quite confused

There is no way to transfer review comments from one PR to another. I've already started reviewing this package in this PR. It should be trivial for you to merge your other branch into this one.

MementoRC replied:
Separated `libsecp256k1` which builds using CMake. This is resolved

MementoRC commented:
Thank you very much for the feedback. I will look at it tomorrow.
I did work on preparing a libsecp256k1 recipe and a coincurve recipe that
uses it as a dependency, if you don't mind taking a brief look at them
It's coincurve_dynamic, still in draft since I was awaiting your feedback
This is the first time I attempt providing a contact recipe, it's very
exciting!

Thank you

On Tue, Oct 3, 2023, 5:50 PM Daniel Ching ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In recipes/coincurve/meta.yaml
> <https://github.com/conda-forge/staged-recipes/pull/23741#discussion_r1344838763>
> :
>
> > +    - autoconf  # [not win]
> +    - automake  # [not win]
> +    - libtool  # [not win]
> +    - m4  # [not win]
> +    - make  # [not win]
> +    - pkg-config  # [not win]
> +    - asn1crypto  # [win]
> +    - m2-bash  # [win]
> +    - m2-autoconf  # [win]
> +    - m2-automake-wrapper  # [win]
> +    - m2-gzip  # [win]
> +    - m2-libtool  # [win]
> +    - m2-m4  # [win]
> +    - m2-pkg-config  # [win]
> +    - m2-sed  # [win]
> +    - m2w64-gcc  # [win]
> +    - m2w64-make  # [win]
> +    - posix  # [win]
> +    - pydflow  # [win]
>
> I would suggest using msvc toolchain instead of m2 if you have the option.
> The m2 conda toolchain isn't maintained well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/conda-forge/staged-recipes/pull/23741#pullrequestreview-1656255771>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWGZEZ64XAXIBONYKFLX7E3X5SJDDAVCNFSM6AAAAAA3YP7NM2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTMNJWGI2TKNZXGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

MementoRC commented:
@carterbox I think I followed the gist of your suggestions and I think it is ready for the next pass at review

carterbox commented:
Our channel strongly prefers dynamic linking over static linking because it reduces resources spent on building, hosting, and delivering packages. There really has to be a compelling reason for static linking on our channel.

Please organize your outputs as follows:

secp256k1: binaries or other CLIs; depends on libsecp256k1
libsecp256k1-dev: Headers defining the functions provided by the library, cmake/pkconfig files; depends on libsecp256k1
~~libsecp256k1-static: Static library~~ allowed for personal use, but discouraged
libsecp256k1: Shared library only

coincurve: Shared python bindings, dynamically linked against libsecp256k1; depends on libsecp256k1

> I was also working on a pybind11 bundle-in to the libsecp256k1, but I hope that once libsecp256k1 is a feedstock, I could add it to it, if I ever complete it

I'm not sure what you mean. The artifacts created in these conda recipes should be as close to the upstream projects as possible. You should not be adding any new functionality or making modifications unless it is for the purpose of getting the artifacts to work within the conda environment / ecosystem. i.e. Adding a new pybind11 wrapper of libsecp256k1 is not allowed unless it has been already merged [upstream](https://github.com/bitcoin-core/secp256k1).

MementoRC commented:
ok

MementoRC commented:
@carterbox 
The changes were implemented. The recipe provides::
   - `libsecp256k1`: Shared library
   - `coincurve`: Shared python bindings

MementoRC commented:
@conda-forge/help-python-c ready for review

carterbox commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,80 @@
+  source_files:
+    - tests/
+  requires:
+    - {{ pin_compatible('libsecp256k1') }}
```
```suggestion
```
Shouldn't be necessary because this package (coincurve) is installed in the tests environment and already depends on libsecp256k1.

MementoRC replied:
Will do

carterbox commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,80 @@
+    This library provides well-tested Python bindings for libsecp256k1, 
+    the heavily optimized C library used by Bitcoin Core for operations 
+    on the elliptic curve secp256k1.
+  license: MIT
```
```suggestion
  license: MIT OR APACHE
```
Use the OR operator to list the identifiers for both licenses. Not sure if APACHE is the correct identifier name.

MementoRC replied:
WIll do

carterbox commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,80 @@
+source:
+  url: https://github.com/ofek/coincurve/archive/{{ commit }}.tar.gz
+  sha256: {{ sha256 }}
+  patches:
+    - update_pyproject.toml.patch
+    - update_setup.py.patch
+    - update_build.py.patch
+    - update_MANIFEST.in.patch
+    - update_coincurve_deprecated.patch
```
Please add a note explaining why these patches are implemented. Some of them seem to be because we want to build libsecp in a separate feedstock (which is fine), but there also seem to be some that change the coincurve API. Probably, the patches that change the coincurve API should be merged upstream.

MementoRC replied:
`coincurve` provides a binding to `libsecp256k1`. However, libsecp256k1 is not a pypi package, so the author bundles it into his pypi package. But, with the requirements of conda-forge, `libsecp256k1` is a conda package, so providing `coincurve` requires modifying (significantly?) the way the package is built. In doing so, a few monior modifications were made to simplify the future updates in `libsecp256k1`

   - setup.py: Major refactoring to remove the no longer needed `'ibsecp256k1` build
   - setup_support.py: No longer needed (consequence of above)
   - build.py: Provision to more readily integrate updates from `libsecp256k1`
   - MANIFEST.in: removal of no longer needed references
   - coincurve_deprecated: Replace usage of deprecated calls
   - pyproject.toml: Simplify setup/pyproject
   
From the learning of conda-forge recipes, it seems that this list of changes disqualifies this PR and the proposed alternative will be to create an upstream that reflect these changes?

carterbox replied:
Patching the build system is not the problem. Downstream users will NOT notice that you moved the metadata definition from setup.py to pyproject.toml. What concerns me is "coincurve_deprecated: Replace usage of deprecated calls". This patch seems to be modifying the coincurve API, something that downstream users of this package will notice.

MementoRC replied:
Agreed. It seemed counter productive to not fix issues in the original package and update to the latest `libsecp256k1` , but this is indeed opposite to the charter of conda-forge.
A proposal could be to revert these API changes and provide a separate package that implements the desired upgrades?

carterbox replied:
You should use the same version of libsecp256k1 that the upstream coincurve package uses. If you really want to use a newer version of libsecp256k1. Work with the maintainers of coincurve to accept your upgrade to the new libsecp256k1 version. Wait until coincurve publishes a new release with your changes before publishing them here.

If coincurve is abandoned, then you can create your own fork of coincurve, but you cannot publish your fork as `coincurve` on this channel; you would need to publish it as `momentorc-coincurve` or change the project name to something new.

MementoRC replied:
Understood. Closing PR until conformance to conda-forge

carterbox reviewed (changes requested):
Looks like you have learned a lot! Your recipes are much better than before.

carterbox commented on recipes/libsecp256k1/meta.yaml:
```diff
@@ -0,0 +1,105 @@
+build:
+  number: 0
+  missing_dso_whitelist:
+    - "*/libc.so.6"
```
Please post the logging message without libc in the missing_dso_whitelist?

carterbox replied:
You should be seeing something like:

```
2023-11-16T18:21:36.0530842Z    INFO (libtomo,lib/libtomo-gridrec.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libc.so.6 found in CDT/compiler package conda-forge::sysroot_linux-64-2.17-h4a8ded7_13
```

i.e. The linker should be linking your objects to libc in our placeholder sysroot package.

MementoRC replied:
The log shows something like this instead:
`Making absolute symlink relative (lib/libsecp256k1.so -> libsecp256k1.so.2 :-> libsecp256k1.so.2.1.0)
  ERROR (libsecp256k1,lib/libsecp256k1.so.2.1.0): /lib64/libc.so.6 not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
`
How to resolve?

carterbox replied:
It's because you put the compiler in requirements/host instead of requirements/build. Build tools go in requirements/build. Deps that are linked to and need to match the target platform when cross-compiling go in requirements/host.

carterbox replied:
I changed deps from pkgconfig (python) to pkg-config (CLI). I'm not sure if that was correct or not.

MementoRC commented on recipes/libsecp256k1/0001_update_pc.in.patch:
```diff
@@ -0,0 +1,25 @@
+-exec_prefix=@exec_prefix@
+-libdir=@libdir@
+-includedir=@includedir@
++prefix=/opt/anaconda1anaconda2anaconda3
++exec_prefix=${prefix}
++libdir=${exec_prefix}/lib
++includedir=${prefix}/include
```
The `/opt/anaconda...` is described a s a placeholder in: https://docs.conda.io/projects/conda-build/en/stable/resources/make-relocatable.html, is it obsolete?

carterbox replied:
I've never used that method of placeholder before! If it's in the docs, it should be fine.

MementoRC commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,80 @@
+  version: {{ version }}
+
+source:
+  url: https://github.com/ofek/coincurve/archive/{{ commit }}.tar.gz
```
Will do

MementoRC commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,80 @@
+    - vc40_runtime
+    - ucrt
+
+requirements:
+  host:
+    - {{ compiler('c') }}
+    - {{ pin_compatible('libsecp256k1') }}
+    - cffi >=1.3.0
+    - pip
+    - pkgconfig
+    - python
+  run:
+    - asn1crypto
+    - cffi >=1.3.0
+    - {{ pin_compatible('libsecp256k1') }}
+    - python
```
Will do

MementoRC commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,80 @@
+    - "*/VCRUNTIME*.dll"  # [win]
+    - "*/ucrtbased.dll"  # [win]
+    - "*/KERNEL32.dll"  # [win]
+  ignore_run_exports:
+    - libsecp256k1
+    - cffi
+    - python
```
During the build, there are warnings that seems to indicate an ill-defined recipe. This will be removed

MementoRC commented on recipes/libsecp256k1/meta.yaml:
```diff
@@ -0,0 +1,105 @@
+    - 0002_update_src_cmake_for_pc.in.patch
+    - 0003_create_pkgconfig_function.patch
+
+build:
+  number: 0
```
Will do

MementoRC commented on recipes/libsecp256k1/meta.yaml:
```diff
@@ -0,0 +1,105 @@
+  version: {{ version }}
+
+source:
+  url: https://github.com/bitcoin-core/secp256k1/archive/{{ commit }}.tar.gz
```
Will do

MementoRC commented:
Close till resolution

carterbox commented:
Please reopen this same PR when you are ready to continue.

MementoRC commented:
@carterbox both recipes should aligned to conda-forge, the name `coincurve` cannot be used based on the previous points.

`coincurve` needs a particular commit of `libsecp256k1` pre-release tag. The way to provide it as a recipe could be to create a recipe based on this one but with a fixed version of `libsecp256k1` based of the commit used in the upstream. Otherwise, there seem to be no suitable way of providing the `coincurve` recipe

carterbox commented:
> coincurve needs a particular commit of libsecp256k1 pre-release tag. The way to provide it could be to create a recipe based on this one but with a fixed version of libsecp256k1 based of the commit used in the upstream. Otherwise, there seems to be no suitable way of providing the coincurve recipe.

@MomentoRC, It looks like you were able to get coincurve migrated to a stable release v0.3.2 (https://github.com/ofek/coincurve/pull/126) of libsecp256k1, so don't you want to start this recipe at v0.3.2, so you can use it for coincurve?

ofek commented on recipes/libsecp256k1-py-bindings/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+    - pip check
+    - pytest -v tests
+
+about:
+  home: https://github.com/ofek/coincurve
+  summary: Cross-platform Python CFFI bindings for libsecp256k1
+  doc_url: https://ofek.dev/coincurve/
+  description: |
+    This library provides well-tested Python bindings for libsecp256k1,
+    the heavily optimized C library used by Bitcoin Core for operations
+    on the elliptic curve secp256k1.
+  license: MIT OR APACHE-2.0
+  license_file:
+    - LICENSE-MIT
+    - LICENSE-APACHE
```
Wait does coincurve need a new release?

carterbox replied:
Yes. I believe that the latest release of coincurve (v18) links to a version of libsekcp256k1 that is too old for @MementoRC 's use case. However, I'm not sure whether even the main branch of coincurve contains a new enough version since that version is v0.3 and this PR contains a fork of coincurve which links to v0.4. Please coordinates amongst yourselves.

MementoRC replied:
@ofek, A short answer is to provide a release with the recent merge so that the conda package could rely upon a release tag rather than a commit. But still, that would not resolve the key issue

Based on my understanding the main issue is that, because it repackages the `libsecp256k1` library from bitcoin-core, `coincurve` cannot be a clean conda package. The solution I propose is to provide the binding package that relies upon a separate conda package (C lib) that provides `libsecp256k1` - However, that package does not exist in Pypi and thus it would require you to create a backward incompatibility with how you deliver your package (binding+C lib)

The conundrum is that, a conda recipe cannot change the upstream resource, so all the patches that I made to provide a conda packaging of coincurve are not acceptable

ofek replied:
Oh, I see what you mean. That's a tough situation. Let me know if I can help.

carterbox replied:
> But still, that would not resolve the key issue

It would resolve the issue because coincurve will [detect and link dynamically to existing system libraries of libsecp256k1](https://github.com/ofek/coincurve/blob/2dfea673835e524ffaeb11dc1c2aecaf23188195/setup_support.py#L84). So repackagaging is not the issue because coincurve can dynamically link to a system provided libsecp256k1 (the static linking / repackaging is already optional).

The issue is whether the latest stable release of coincurve can link to a stable release of libsecp256k1.

If @ofek makes a new release for coincurve v19 that is ABI compatible with libsecp256k1 v0.3.2 then both coincurve and libsecp256k1 will be build from stable releases. That should solve the problem as far as this channel is concerned.

carterbox replied:
@MementoRC, please see comment above.

MementoRC replied:
@carterbox but how would libsecp256k1 be made available in a non-conda environment? Or are you saying that it is not necessary to remove the repackaging, but create a conda package that requires the `libsecp256k1` conda package?
As-is, the _has_system_lib will not work as the path is hard-coded
@ofek, I plan to look into providing a PR for version 0.4.0

carterbox replied:
> how would libsecp256k1 be made available in a non-conda environment?

That's irrelevant? Conda packages are only for conda environments.

> Is not necessary to remove the repackaging, but create a conda package that requires the libsecp256k1 conda package?

Yes. I'm pretty sure that coincurve [uses externally provided libsecp256k1 by default](https://github.com/ofek/coincurve/blob/1dd0675add2da67ad9e6acfc47eed041066f2ea1/setup.py#L134) and only revends if it cannot find an externally provided libsecp256k1. You just need to add the libsec256k1 package to the host dependencies of the coincurve recipe and ensure that the coincurve setup script can find it correctly; this probably requires no/very few patches.

You can probably even use v0.4.0 of libsecp256k1 because it is ABI compatible with v0.3.x.

carterbox commented:
I have updated the libsecp256k1 recipe to ship the dynamic libraries separately from the headers/import/cmake/pc files. This makes the run_export smaller and better at expressing compatibility because the SONAME (ABI version) is used to determine compatibility instead of the API version; for libsecp256k1, the developers are versioning the ABI/API separately (which is good).

carterbox commented:
These recipes are very close to being ready to merge, just need to clean up the metadata changes from the conversion from coincurve to libsecp256k1-py-bindings.

MementoRC commented on recipes/libsecp256k1-py-bindings/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+{% set name = "libsecp256k1-py-bindings" %}
+{% set version = "0.4.0" %}
```
@carterbox Oh right. Will do

MementoRC replied:
Done

ofek commented:
I'll release a new version of coincurve tonight

MementoRC commented:
@carterbox , that is a awesome peice of conda voodoo right there ... thank you for the update.

MementoRC commented:
@ofek, the major issue is thatI don't see how to provide `coincurve` as a conda package since it repackages the C library. It also refers to a specific commit, which I understand from your standpoint, but is an issue for a conda-recipe
@carterbox Do you see how we could package coincurve? My interest from the start was to provide coincurve (which is widely used, not libsecp256k1-py-bindings, which may never be used by anyone ...)

MementoRC commented:
@carterbox I separated the upstream from coincurve to reduce the confusion and open a simpler way to a future resolution on providing a compliant `coincurve` conda package

MementoRC commented:
@carterbox, I am adding `coincurve` recipe again. It is WIP since it will depend upon @ofek accepting the PR for update to SECP256K1 v0.4.0 that I submitted. I am using my fork for which i have made a release with the PR I proposed.

MementoRC commented:
@carterbox Ok, it works correctly for linux:
```
  running egg_info
  creating coincurve.egg-info
  writing coincurve.egg-info/PKG-INFO
  writing dependency_links to coincurve.egg-info/dependency_links.txt
  writing requirements to coincurve.egg-info/requires.txt
  writing top-level names to coincurve.egg-info/top_level.txt
  writing manifest file 'coincurve.egg-info/SOURCES.txt'
  reading manifest file 'coincurve.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  no previously-included directories found matching '/*'
  warning: no directories found matching 'libsecp256k1'
  adding license file 'LICENSE-APACHE'
  adding license file 'LICENSE-MIT'
  writing manifest file 'coincurve.egg-info/SOURCES.txt'
  copying coincurve/py.typed -> build/lib.linux-x86_64-cpython-311/coincurve
  running build_ext
  generating cffi module 'build/temp.linux-x86_64-cpython-311/_libsecp256k1.c'
  creating build/temp.linux-x86_64-cpython-311
  building '_libsecp256k1' extension
  creating build/temp.linux-x86_64-cpython-311/build
  creating build/temp.linux-x86_64-cpython-311/build/temp.linux-x86_64-cpython-311
  /home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_build_env/bin/x86_64-conda-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/include -fPIC -O2 -isystem /home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/include -fdebug-prefix-map=/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/work=/usr/local/src/conda/coincurve-19.0.0 -fdebug-prefix-map=/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p=/usr/local/src/conda-prefix -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/include -fPIC -I/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/include/python3.11 -c build/temp.linux-x86_64-cpython-311/_libsecp256k1.c -o build/temp.linux-x86_64-cpython-311/build/temp.linux-x86_64-cpython-311/_libsecp256k1.o
  /home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_build_env/bin/x86_64-conda-linux-gnu-cc -shared -Wl,--allow-shlib-undefined -Wl,-rpath,/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib -Wl,-rpath-link,/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib -L/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib -Wl,--allow-shlib-undefined -Wl,-rpath,/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib -Wl,-rpath-link,/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib -L/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib -Wl,-rpath-link,/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib -L/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/include -fdebug-prefix-map=/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/work=/usr/local/src/conda/coincurve-19.0.0 -fdebug-prefix-map=/home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p=/usr/local/src/conda-prefix -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/conda/staged-recipes/build_artifacts/coincurve_1703094594067/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/include build/temp.linux-x86_64-cpython-311/build/temp.linux-x86_64-cpython-311/_libsecp256k1.o -lsecp256k1 -o build/lib.linux-x86_64-cpython-311/coincurve/_libsecp256k1.cpython-311-x86_64-linux-gnu.so
  installing to build/bdist.linux-x86_64/wheel
```

But of course, windows will never give me a pass and get off my back ... dang, what a guttural hatred I have for it

ofek commented:
FYI I cross compile to Windows because I didn't know how to do it properly

carterbox commented:
@MementoRC Do you have access to a Windows machine? If not, I have some time this weekend that I can take a look.

MementoRC commented:
@carterbox well, it's most of the time my own mistake, as in this case hard-coding a `/` in setup.py, so it downloaded secp256k1 and added a full path in SOURCES.txt (I don't know why `manifest_maker` adds full path then complains that there is a full path!)
Is there a way to rerun the `checks`? i updated the coincurve release

@ofek, thanks, but the issue is not with the compilation, in fact this recipe must not download and compile `libsecp256k1` but use the conda package

carterbox commented:
@MementoRC, adding a new commit or closing/reopening the PR restarts the checks.

MementoRC commented:
@ofek, `coincurve` builds: build/lib.linux-x86_64-cpython-310/coincurve/_libsecp256k1.cpython-310-x86_64-linux-gnu.so (as well as py-bindings)

ofek commented:
Awesome!!!

MementoRC commented:
@carterbox @ofek I think we finally got it cleaned-up

carterbox commented on recipes/libsecp256k1-py-bindings/meta.yaml:
```diff
@@ -0,0 +1,59 @@
+    - pytest -v tests
+
+about:
+  home: https://github.com/MementoRC/{{ name }}
```
Please don't use variables in the about section urls because then you can't right click to open them directly.

MementoRC replied:
Done

carterbox commented on recipes/coincurve/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+  version: {{ version }}
+
+source:
+  url: https://github.com/MementoRC/coincurve/archive/refs/tags/pre-v{{ version }}.tar.gz
+  sha256: 6d6045f0b6662d7f16c319ba7637206a5c3c6673e1e32d5a2f3dfc2b37888405
```
This recipe needs to use a tagged release from ofek/coincurve before it is merged.

MementoRC replied:
Updated

carterbox reviewed (changes requested):
All the builds seem to be linking correctly, have the correct dependency relationships, and contain the appropriate artifacts, so we're just waiting on a tagged release for the latest coincurve version.

ofek commented:
https://github.com/ofek/coincurve/releases/tag/v19.0.0

MementoRC commented:
@carterbox It should be ready for your review

MementoRC commented:
@ofek, well, yes, but it does not seem related to your release. The 3 are failing for different reasons too, I think it is a recipe issue, I got these before in other receipes, but I forgot how to resolve them :/

carterbox commented:
```
['  ERROR (libsecp256k1,lib/libsecp256k1.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libc.so.6 not found in any CDT/compiler package, nor the whitelist?!']
```
This is not an error from your package. It is a problem with our compiler packages. Please update this branch to see if the error resolves.

carterbox commented:
@conda-forge/core I believe there is an error with the link checker in conda-build version `3.28.3`. When building the `libsecp256k1` recipe from this pull request, we get a missing dependency error as follows:

```
['  ERROR (libsecp256k1,lib/libsecp256k1.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libc.so.6 not found in any CDT/compiler package, nor the whitelist?!']
```
It seems that the link checker cannot find `libc.so.6` for the `libsecp256k1` output even the compiler jinja template *is* included in the build requirements section for that output.

The same requirements are used for the `libsecp256k1-2` output, but the link checker can find `libc` there. 

I was able to recproduce the issue locally, but *only* with conda-build `3.28.3`.

```
conda build -m .ci_support/linux64.yaml recipes/libsecp256k1 --error-overlinking
```
Earlier versions of conda-build including `3.26`, `3.27` did not check the links of `libsecp256k1.so` (because it is a symlink?), and so do not raise the error. `3.28.2` checks the links but is able to find `libc`.

carterbox commented:
@MementoRC, using a different version of conda-build seems to allows the builds to succeed. Please allow me more time to now review the log output and discuss what to do about the broken conda-build version.

MementoRC commented:
@carterbox Cool. Note that when I install secp256k1 from source, it creates a `so.2.1.2`, Let me know how it goes. Thank you

carterbox reviewed (approved):
Build logs LGTM.

> Note that when I install secp256k1 from source, it creates a so.2.1.2, Let me know how it goes. Thank you

The build logs here clearly show that the so version is 2.1.1. Are you building locally with the release config? Maybe they automatically bump the patch number when they detect it's not a release build.

carterbox commented:
I am going to update my commit message that prevents conda-build 3.28.3 from being used, then merge this PR. You will probably need to patch the feedstock to prevent conda-build 3.28.3 from being used in `.scripts/build_steps.sh`

MementoRC commented:
@carterbox  Thank you. This was a long journey

jakirkham commented:
It might be possible to pin `conda-build` [using `remote_ci_setup` in `conda-forge.yml`]( https://conda-forge.org/docs/maintainer/conda_forge_yml.html#remote-ci-setup )

Note that `conda-forge-ci-setup` will need to be in that YAML `list` as well otherwise there will be build errors

Also note would [re-render]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) after making these changes
